stdlib/public/core/Dictionary.swift

```diff
@@ -1541,9 +1541,17 @@ internal struct _DictionaryAnyHashableBox<Key: Hashable, Value: Hashable>
     return _value
   }
 
+  internal var _canonicalBox: _AnyHashableBox {
+    return _DictionaryAnyHashableBox<AnyHashable, AnyHashable>(_canonical)
+  }
+
   internal func _isEqual(to other: _AnyHashableBox) -> Bool? {
-    guard let other = other._asDictionary() else { return nil }
-    return _canonical == other
+    guard
+      let other = other as? _DictionaryAnyHashableBox<AnyHashable, AnyHashable>
```

Comment:
Wait, shouldn't you be downcasting `other._canonicalBox` to `_DictionaryAnyHashableBox<AnyHashable, AnyHashable>` here?

Reply:
`_isEqual` is guaranteed to be only called on canonical boxes.

(I remember adding this guarantee as an optimization, but now I can't see how this is better compared to canonicalizing `self` and `other` right here. Hm...)

Reply:
Ah, I added the guarantee because canonicalization sometimes results in one of several unrelated types (e.g., a canonicalized `Double` box could be either `_DoubleAnyHashableBox` or `_IntegerAnyHashableBox<Int64/UInt64>`), and it was unwieldy to manage those directly in `_isEqual`.